File: packages/completion/engine-jvm/src/main/kotlin/io/autocomplete/skip/ContextualSkip.kt
package io.autocomplete.skip

/**
 * Contextual skip rules adapted from Kilo classic autocomplete behavior.
 */
object ContextualSkip {
    private val languageTerminators: Map<String, List<String>> =
        mapOf(
            "javascript" to listOf(";", "}", ")"),
            "javascriptreact" to listOf(";", "}", ")"),
            "typescript" to listOf(";", "}", ")"),
            "typescriptreact" to listOf(";", "}", ")"),
            "java" to listOf(";", "}", ")"),
            "kotlin" to listOf(";", "}", ")"),
            "scala" to listOf(";", "}", ")"),
            "c" to listOf(";", "}", ")"),
            "cpp" to listOf(";", "}", ")"),
            "csharp" to listOf(";", "}", ")"),
            "go" to listOf(";", "}", ")"),
            "rust" to listOf(";", "}", ")"),
            "php" to listOf(";", "}", ")"),
            "dart" to listOf(";", "}", ")"),
            "css" to listOf(";", "}", ")"),
            "scss" to listOf(";", "}", ")"),
            "less" to listOf(";", "}", ")"),
            "json" to listOf(";", "}", ")"),
            "jsonc" to listOf(";", "}", ")"),
            "python" to listOf(")", "]", "}"),
            "ruby" to listOf(")", "]", "}", "end"),
            "shellscript" to listOf(";", "fi", "done", "esac"),
            "bash" to listOf(";", "fi", "done", "esac"),
            "zsh" to listOf(";", "fi", "done", "esac"),
            "sh" to listOf(";", "fi", "done", "esac"),
            "sql" to listOf(";"),
            "mysql" to listOf(";"),
            "postgresql" to listOf(";"),
        )

    private val defaultTerminators = listOf(";", "}", ")")

    fun shouldSkip(
        prefix: String,
        suffix: String,
        languageId: String,
    ): Boolean {
        if (isMidWord(prefix, suffix)) return true
        if (endsWithTerminator(prefix, languageId)) return true
        return false
    }

    fun isMidWord(
        prefix: String,
        suffix: String,
    ): Boolean {
        if (prefix.isEmpty()) return false
        val before = prefix.last()
        val after = suffix.firstOrNull()
        val word = { c: Char -> c.isLetterOrDigit() || c == '_' }
        if (!word(before)) return false
        if (after != null && word(after)) return true
        // typing in the middle of an identifier-like token without whitespace break
        val line = prefix.substringAfterLast('\n')
        if (line.isEmpty()) return false
        val trailing = line.takeLastWhile { word(it) }
        return trailing.length >= 24
    }

    fun endsWithTerminator(
        prefix: String,
        languageId: String,
    ): Boolean {
        val trimmed = prefix.trimEnd()
        if (trimmed.isEmpty()) return false
        val terms = languageTerminators[languageId.lowercase()] ?: defaultTerminators
        return terms.any { term ->
            if (term.length == 1) {
                trimmed.endsWith(term)
            } else {
                val re = Regex("""(?:^|[\s;{}()])""" + Regex.escape(term) + """\s*$""")
                re.containsMatchIn(trimmed)
            }
        }
    }
}

File: packages/completion/engine-ts/src/contextualSkip.ts
const languageTerminators: Record<string, string[]> = {
  javascript: [";", "}", ")"],
  javascriptreact: [";", "}", ")"],
  typescript: [";", "}", ")"],
  typescriptreact: [";", "}", ")"],
  java: [";", "}", ")"],
  kotlin: [";", "}", ")"],
  scala: [";", "}", ")"],
  c: [";", "}", ")"],
  cpp: [";", "}", ")"],
  csharp: [";", "}", ")"],
  go: [";", "}", ")"],
  rust: [";", "}", ")"],
  php: [";", "}", ")"],
  dart: [";", "}", ")"],
  css: [";", "}", ")"],
  scss: [";", "}", ")"],
  less: [";", "}", ")"],
  json: [";", "}", ")"],
  jsonc: [";", "}", ")"],
  python: [")", "]", "}"],
  ruby: [")", "]", "}", "end"],
  shellscript: [";", "fi", "done", "esac"],
  bash: [";", "fi", "done", "esac"],
  zsh: [";", "fi", "done", "esac"],
  sh: [";", "fi", "done", "esac"],
  sql: [";"],
  mysql: [";"],
  postgresql: [";"],
};

const defaultTerminators = [";", "}", ")"];

function isWordChar(c: string): boolean {
  return /[A-Za-z0-9_]/.test(c);
}

export function shouldSkip(prefix: string, suffix: string, languageId: string): boolean {
  if (isMidWord(prefix, suffix)) return true;
  if (endsWithTerminator(prefix, languageId)) return true;
  return false;
}

export function isMidWord(prefix: string, suffix: string): boolean {
  if (!prefix) return false;
  const before = prefix[prefix.length - 1]!;
  const after = suffix[0];
  if (!isWordChar(before)) return false;
  if (after && isWordChar(after)) return true;
  const line = prefix.slice(prefix.lastIndexOf("\n") + 1);
  if (!line) return false;
  let trailing = "";
  for (let i = line.length - 1; i >= 0; i--) {
    if (!isWordChar(line[i]!)) break;
    trailing = line[i]! + trailing;
  }
  return trailing.length >= 24;
}

export function endsWithTerminator(prefix: string, languageId: string): boolean {
  const trimmed = prefix.replace(/\s+$/, "");
  if (!trimmed) return false;
  const terms = languageTerminators[languageId.toLowerCase()] ?? defaultTerminators;
  return terms.some((term) => {
    if (term.length === 1) return trimmed.endsWith(term);
    const re = new RegExp(`(?:^|[\\s;{}()])${escapeRegExp(term)}\\s*$`);
    return re.test(trimmed);
  });
}

function escapeRegExp(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}
